fix: Don't trust the length reported by the ExactSizeIterator trait

Marwes opened the pull request:
Since the `ExactSizeIterator` trait is not an unsafe trait we can't trust the length it reports when creating the slice. Doing so would make it possible to pass an iterator to `alloc_extend` which is larger than the length it reports, causing the arena to be reallocated.

As a bonus `alloc_extend` now only requires the `I: IntoIterator<Item = T>` trait

cc #7

pczarn commented on src/lib.rs:
```diff
@@ -64,18 +64,42 @@ impl<T> Arena<T> {
+            next_item_index = chunks.current.len();
+            let mut i = 0;
+            while let Some(elem) = iter.next() {
+                if i == iter_min_len {
```
I think you could compare `iter_min_len` or `iter_max_len` here, whichever fits in the chunk's remaining space, right?

Marwes replied:
👍 I will make it compare against the current capacity instead which should be clearest I think.

Marwes replied:
By comparing capacity == len it should handle an erroneous size_hint a bit better. If size_hint().0 is less than the actual size but the actual iterator still fits into the current chunk it will not reallocate unnecessarily but will use the current check instead.

pczarn commented on src/lib.rs:
```diff
@@ -64,18 +64,42 @@ impl<T> Arena<T> {
+                    // The iterator were larger than its `size_hint` indicated.
+                    let chunks = &mut *chunks;
+                    // Create a new chunk which we can freely push the entire iterator into
+                    chunks.reserve(1);
```
Perhaps inlining the code by hand or making a new method similar to `reserve` would make the intention clearer. Every time I see this line, I find it hard to imagine what it does.

Marwes replied:
Not really sure how inlining or creating a new method would help. `reserve` does exactly what we want in this case. It _reserves_ an additional chunk which has room for at least one more element (I should actually be passing `iter_min_len + 1` instead since `reserve` takes the minimum size of the next chunk, that's only a matter of avoiding reallocations though).

pczarn replied:
Actually, it reserves an additional chunk with twice as much capacity as the current chunk

Marwes replied:
That's why I said _at least one more element_ :). Having each chunk grow by 2x is already part of and expected of the arena so I haven't changed anything about that.

pczarn replied:
Then it's alright

pczarn commented on src/lib.rs:
```diff
@@ -64,18 +64,42 @@ impl<T> Arena<T> {
+            let mut i = 0;
+            while let Some(elem) = iter.next() {
+                if i == iter_min_len {
+                    // The iterator were larger than its `size_hint` indicated.
```
'the iterator was'

Marwes replied:
Fixed

Marwes commented:
Forgot to add the element which caused the overflow to the fresh chunk so I fixed that as well and added an extra assert at the end of the added test which checks that all elements are added in the wrong size case.

pczarn commented:
Excellent solution. Thanks

Marwes commented:
:) Since there does not seem to be much activity on this crate a release of this would be nice (last release was almost a year ago!), otherwise this might never get to crates.io.

pczarn commented:
Oh right, I released it

Marwes commented:
Excellent, thanks!
